# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/CanvasRenderer.cpp
# Commit: 05896dbeb434
# Full Hash: 05896dbeb434d6f3ca9324f01eb67020d932f50f
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-06-15 21:48:38
# Regressor Bug: 1632249
# File Overlap Count: 1
# Description:
#   Bug 1632249 - Determine webgl's layers::TextureType in Present. r=lsalzman
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D79617
# ==============================================================================

diff -r 8439f89e170b -r 05896dbeb434 gfx/layers/CanvasRenderer.cpp
--- a/gfx/layers/CanvasRenderer.cpp	Mon Jun 15 18:26:12 2020 +0000
+++ b/gfx/layers/CanvasRenderer.cpp	Mon Jun 15 18:26:14 2020 +0000
@@ -6,6 +6,9 @@
 
 #include "CanvasRenderer.h"
 
+#include "BuildConstants.h"
+#include "ipc/KnowsCompositor.h"
+#include "mozilla/StaticPrefs_webgl.h"
 #include "nsICanvasRenderingContextInternal.h"
 #include "PersistentBufferProvider.h"
 #include "WebGLTypes.h"
@@ -79,5 +82,52 @@
   context->OnDidPaintTransaction();
 }
 
+TextureType TexTypeForWebgl(KnowsCompositor* const knowsCompositor) {
+  if (!knowsCompositor) return TextureType::Unknown;
+  const auto layersBackend = knowsCompositor->GetCompositorBackendType();
+
+  switch (layersBackend) {
+    case LayersBackend::LAYERS_NONE:
+      MOZ_CRASH("Unexpected LayersBackend::LAYERS_NONE");
+    case LayersBackend::LAYERS_CLIENT:
+      MOZ_CRASH("Unexpected LayersBackend::LAYERS_CLIENT");
+    case LayersBackend::LAYERS_LAST:
+      MOZ_CRASH("Unexpected LayersBackend::LAYERS_LAST");
+
+    case LayersBackend::LAYERS_BASIC:
+      return TextureType::Unknown;
+
+    case LayersBackend::LAYERS_D3D11:
+    case LayersBackend::LAYERS_OPENGL:
+    case LayersBackend::LAYERS_WR:
+      break;
+  }
+
+  if (kIsWindows) {
+    if (knowsCompositor->SupportsD3D11()) {
+      return TextureType::D3D11;
+    }
+    return TextureType::Unknown;
+  }
+  if (kIsMacOS) {
+    return TextureType::MacIOSurface;
+  }
+  if (kIsX11) {
+    return TextureType::X11;
+  }
+  if (kIsWayland) {
+    if (gfxPlatform::GetPlatform()->UseWaylandDMABufWebGL()) {
+      return TextureType::WaylandDMABUF;
+    }
+  }
+  if (kIsAndroid) {
+    if (StaticPrefs::webgl_enable_surface_texture()) {
+      return TextureType::AndroidNativeWindow;
+    }
+  }
+
+  return TextureType::Unknown;
+}
+
 }  // namespace layers
 }  // namespace mozilla